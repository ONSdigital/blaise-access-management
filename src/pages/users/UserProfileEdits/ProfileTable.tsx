import React from "react";
import { User } from "blaise-api-node-client";
import { Link } from "react-router-dom";
import { GetUserResponse } from "../../../Interfaces/usersPage";

export default function ProfileTable({ currentUser, viewedUserDetails }: {currentUser: User, viewedUserDetails: GetUserResponse}) {
    const { name = "", role = "", defaultServerPark = "", serverParks = [] } = viewedUserDetails?.data ?? {};
    const { name: currentUsername } = currentUser;

    return (
        <main id="main-content" className="ons-page__main ons-u-mt-no">
            <div className="ons-summary">
                <div id="personal-details" className="ons-summary__group">
                    <h2 className="ons-summary__group-title">User profile</h2>
                    <dl className="ons-summary__items">
                        <div id="name-row" className="ons-summary__item">
                            <div className="ons-summary__row ons-summary__row--has-values" id="name">
                                <dt className="ons-summary__item-title">
                                    <div className="ons-summary__item--text">Name</div>
                                </dt>
                                <dd className="ons-summary__values">
                                    <span className="ons-summary__text">{name ? name : "Not found"}</span>
                                </dd>
                                <dd className="ons-summary__actions">
                                </dd>
                            </div>
                        </div>
                        <div id="password-row" className="ons-summary__item">
                            <div className="ons-summary__row ons-summary__row--has-values" id="password">
                                <dt className="ons-summary__item-title">
                                    <div className="ons-summary__item--text">Password</div>
                                </dt>
                                <dd className="ons-summary__values">
                                    <span className="ons-summary__text">{name ? "••••••••" : "Not found"}</span>
                                </dd>
                                <dd className="ons-summary__actions">
                                    {
                                        (
                                            name === currentUsername ?
                                                "Current user" :
                                                <Link to={"/users/changepassword/" + name} state={{ currentUser }} className="ons-summary__button">
                                                    <span className="ons-summary__button-text"
                                                        aria-hidden="true">
                                                Change
                                                    </span>
                                                    <span className="ons-u-vh">Change password</span>
                                                </Link>
                                        )
                                    }
                                </dd>
                            </div>
                        </div>
                    </dl>
                </div>
                <div id="roles-and-permissions" className="ons-summary__group">
                    <h2 className="ons-summary__group-title">Roles and permissions</h2>
                    <div id="role-row" className="ons-summary__item">
                        <div className="ons-summary__row ons-summary__row--has-values" id="role">
                            <dt className="ons-summary__item-title">
                                <div className="ons-summary__item--text">Role</div>
                            </dt>
                            <dd className="ons-summary__values">
                                <span className="ons-summary__text">{role ? role : "Not found"}</span>
                            </dd>
                            <dd className="ons-summary__actions">
                                <Link to={"/users/change-role/" + name} state={{ currentUser, viewedUserDetails }} className="ons-summary__button">
                                    <span className="ons-summary__button-text"
                                        aria-hidden="true">
                                                Change
                                    </span>
                                    <span className="ons-u-vh">Change role</span>
                                </Link>
                            </dd>
                        </div>
                    </div>
                    <div id="default-server-park-row" className="ons-summary__item">
                        <div className="ons-summary__row ons-summary__row--has-values" id="default-server-park">
                            <dt className="ons-summary__item-title">
                                <div className="ons-summary__item--text">Default Server Park</div>
                            </dt>
                            <dd className="ons-summary__values">
                                <span className="ons-summary__text">{defaultServerPark ? defaultServerPark : "Not found"}</span>
                            </dd>
                            <dd className="ons-summary__actions">
                            </dd>
                        </div>
                    </div>
                    <div id="server-parks-row" className="ons-summary__item">
                        <div className="ons-summary__row ons-summary__row--has-values" id="server-parks">
                            <dt className="ons-summary__item-title">
                                <div className="ons-summary__item--text">Server Parks</div>
                            </dt>
                            <dd className="ons-summary__values">
                                <span className="ons-summary__text">{serverParks ? serverParks.join(", ") : "Not found"}</span>
                            </dd>
                            <dd className="ons-summary__actions">
                            </dd>
                        </div>
                    </div>
                </div>
                <div id="actions" className="ons-summary__group">
                    <h2 className="ons-summary__group-title">Actions</h2>
                    <div id="actions-row">
                        {
                            (
                                name === currentUsername ?
                                    "Cannot delete current user" :
                                    <Link to={"/users/delete/" + name} className="ons-summary__button">
                                        <span className="ons-summary__button-text"
                                            aria-hidden="true">
                                                Delete
                                        </span>
                                        <span className="ons-u-vh">Delete user</span>
                                    </Link>
                            )
                        }
                    </div>
                </div>
            </div>
        </main>
    );
}