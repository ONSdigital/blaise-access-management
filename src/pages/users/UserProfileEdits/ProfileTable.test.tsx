import React from "react";
import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";
import ProfileTable from "./ProfileTable";

const currentUser = {
    name: "CurrentUser",
    role: "DST",
    defaultServerPark: "gusty",
    serverParks: ["gusty"]
};

const viewedUserDetails = {
    data: {
        name: "John Doe",
        role: "IPS Manager",
        defaultServerPark: "gusty",
        serverParks: ["gusty", "cma"]
    },
    status: 200,
    message: "Successfully fetched user details for John Doe"
};

describe("ProfileTable", () => {
    it("matches snapshot", () => {
        const { asFragment } = render(
            <Router>
                <ProfileTable currentUser={currentUser} viewedUserDetails={viewedUserDetails} />
            </Router>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("renders user details correctly", async () => {
        render(
            <Router>
                <ProfileTable currentUser={currentUser} viewedUserDetails={viewedUserDetails} />
            </Router>
        );

        expect(screen.getByText("John Doe")).toBeVisible();
        expect(screen.getByText("IPS Manager")).toBeVisible();
        expect(screen.getByText("gusty")).toBeVisible();
        expect(screen.getByText("gusty, cma")).toBeVisible();

        expect(await screen.findByText("Delete")).toBeVisible();
        const changeButtons = await screen.findAllByText("Change");
        changeButtons.forEach(button => {
            expect(button).toBeVisible();
        });
    });

    it("displays \"Not found\" for missing user details", () => {
        const missingDetails = {
            data: {
                name: "",
                role: "",
                defaultServerPark: "",
                serverParks: []
            },
            status: 500,
            message: "User not found"
        };

        render(
            <Router>
                <ProfileTable currentUser={currentUser} viewedUserDetails={missingDetails} />
            </Router>
        );

        expect(screen.getAllByText("Not found").length).toBeGreaterThan(0);
    });
});