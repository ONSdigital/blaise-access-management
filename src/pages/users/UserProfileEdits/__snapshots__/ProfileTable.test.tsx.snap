// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileTable matches snapshot 1`] = `
<DocumentFragment>
  <main
    class="ons-page__main ons-u-mt-no"
    id="main-content"
  >
    <div
      class="ons-summary"
    >
      <div
        class="ons-summary__group"
        id="personal-details"
      >
        <h2
          class="ons-summary__group-title"
        >
          User profile
        </h2>
        <dl
          class="ons-summary__items"
        >
          <div
            class="ons-summary__item"
            id="name-row"
          >
            <div
              class="ons-summary__row ons-summary__row--has-values"
              id="name"
            >
              <dt
                class="ons-summary__item-title"
              >
                <div
                  class="ons-summary__item--text"
                >
                  Name
                </div>
              </dt>
              <dd
                class="ons-summary__values"
              >
                <span
                  class="ons-summary__text"
                >
                  John Doe
                </span>
              </dd>
              <dd
                class="ons-summary__actions"
              />
            </div>
          </div>
          <div
            class="ons-summary__item"
            id="password-row"
          >
            <div
              class="ons-summary__row ons-summary__row--has-values"
              id="password"
            >
              <dt
                class="ons-summary__item-title"
              >
                <div
                  class="ons-summary__item--text"
                >
                  Password
                </div>
              </dt>
              <dd
                class="ons-summary__values"
              >
                <span
                  class="ons-summary__text"
                >
                  ••••••••
                </span>
              </dd>
              <dd
                class="ons-summary__actions"
              >
                <a
                  class="ons-summary__button"
                  href="/users/changepassword/John Doe"
                >
                  <span
                    aria-hidden="true"
                    class="ons-summary__button-text"
                  >
                    Change
                  </span>
                  <span
                    class="ons-u-vh"
                  >
                    Change password
                  </span>
                </a>
              </dd>
            </div>
          </div>
        </dl>
      </div>
      <div
        class="ons-summary__group"
        id="roles-and-permissions"
      >
        <h2
          class="ons-summary__group-title"
        >
          Roles and permissions
        </h2>
        <div
          class="ons-summary__item"
          id="role-row"
        >
          <div
            class="ons-summary__row ons-summary__row--has-values"
            id="role"
          >
            <dt
              class="ons-summary__item-title"
            >
              <div
                class="ons-summary__item--text"
              >
                Role
              </div>
            </dt>
            <dd
              class="ons-summary__values"
            >
              <span
                class="ons-summary__text"
              >
                IPS Manager
              </span>
            </dd>
            <dd
              class="ons-summary__actions"
            >
              <a
                class="ons-summary__button"
                href="/users/changerole/John Doe"
              >
                <span
                  aria-hidden="true"
                  class="ons-summary__button-text"
                >
                  Change
                </span>
                <span
                  class="ons-u-vh"
                >
                  Change role
                </span>
              </a>
            </dd>
          </div>
        </div>
        <div
          class="ons-summary__item"
          id="default-server-park-row"
        >
          <div
            class="ons-summary__row ons-summary__row--has-values"
            id="default-server-park"
          >
            <dt
              class="ons-summary__item-title"
            >
              <div
                class="ons-summary__item--text"
              >
                Default Server Park
              </div>
            </dt>
            <dd
              class="ons-summary__values"
            >
              <span
                class="ons-summary__text"
              >
                gusty
              </span>
            </dd>
            <dd
              class="ons-summary__actions"
            />
          </div>
        </div>
        <div
          class="ons-summary__item"
          id="server-parks-row"
        >
          <div
            class="ons-summary__row ons-summary__row--has-values"
            id="server-parks"
          >
            <dt
              class="ons-summary__item-title"
            >
              <div
                class="ons-summary__item--text"
              >
                Server Parks
              </div>
            </dt>
            <dd
              class="ons-summary__values"
            >
              <span
                class="ons-summary__text"
              >
                gusty, cma
              </span>
            </dd>
            <dd
              class="ons-summary__actions"
            />
          </div>
        </div>
      </div>
      <div
        class="ons-summary__group"
        id="actions"
      >
        <h2
          class="ons-summary__group-title"
        >
          Actions
        </h2>
        <div
          id="actions-row"
        >
          <a
            class="ons-summary__button"
            href="/users/delete/John Doe"
          >
            <span
              aria-hidden="true"
              class="ons-summary__button-text"
            >
              Delete
            </span>
            <span
              class="ons-u-vh"
            >
              Delete user
            </span>
          </a>
        </div>
      </div>
    </div>
  </main>
</DocumentFragment>
`;
